Feature/test cases util

jkshj21 opened the pull request:
Created the test cases util class that have the following functions

- toggle_webhooks_test_cases : Enable/disable webhooks for the given test case ids. It enables or disables webhooks in each conversation turn. it returns the list of the updated test cases. It gives options to users to either create, update or export/import the test cases.

- toggle_sentiment_test_cases: Enable/disable sentiment analysis for the given test case ids. It enables or disables sentiment analysis in each conversation turn. it returns the list of the updated test cases. It gives options to users to either create, update or export/import the test cases.

- migrate_test_cases: The purpose of this method is to create a new set of the test cases from (old) a source agent to (new) a target agent. When the new agent (target) is created by the copy util, UUIDs of flows, pages, intents, etc are newly generated. Therefore, the discrepancies of the uuids create conflict when importing the test cases from a source to a target agent. To prevent this, this method helps to migrate the test cases by replacing the UUIDs that are relevant to a target agent. The test cases are only migratable if the display names of flows, pages, intents, etc have the matching set in both source and target agents.

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,552 @@
+
+        if agent_id:
+            self.agent_id = agent_id
+            self.client_options = self._set_region(self.agent_id)
```
@jkshj21 remove these. they're unused in this class

jkshj21 replied:
Removed

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,552 @@
+            self.agent_id = agent_id
+            self.client_options = self._set_region(self.agent_id)
+
+        if test_case_id:
```
Remove this section, none of these are used in this class.

jkshj21 replied:
Removed

jkshj21 commented:
Ok, @kmaphoenix 

Thank you for your feedback 
I will work on it

jkshj21 commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,552 @@
+
+        return page_id
+
+    def _get_intent_id_by_intent_name(
```
Yes, I agree... I think it would be efficient if I store both intents_map, pages_map, flows_map (standard and reverse) in common_config dict. So, the method calls them once if necessary. 

--Flows map ---
`    def _get_flow_id_by_flow_name(
        self,
        flow_name: str,
        flows_map: Dict[str, str]) -> str:
        """Attempt to get the Flow id from the Flows Map
            by the flow display name."""

        **flows_map = {name: id for id, name in flows_map.items()}**
        flow_id = flows_map.get(flow_name, None)`

--Pages map---
`elif page_map:
            **page_map = {name: id for id, name in page_map.items()}**
            page_id = page_map.get(page_name, None)`

jkshj21 replied:
`    def _get_commons_config(
        self,
        agent_id: str) -> Dict[str, Dict[str, str]]:
        """This method map flows, pages, and intents
            are in the dictionary format
          Arg:
            agent_id: The formatted CX Agent ID
          Returns:
            a dictionary of flows, pages, and intents
        """
        commons_config = {}
        dfcx_flows = flows.Flows(creds=self.creds, agent_id=agent_id)
        dfcx_pages = pages.Pages(creds=self.creds)
        dfcx_intents = intents.Intents(creds=self.creds, agent_id=agent_id)
        dfcx_intents_map = dfcx_intents.get_intents_map(agent_id=agent_id)
        dfcx_flows_map = dfcx_flows.get_flows_map(agent_id=agent_id)
        dfcx_pages_map = {}
        for flow_id in dfcx_flows_map.keys():
            dfcx_pages_map[flow_id] = (
                dfcx_pages.get_pages_map(
                  flow_id=flow_id
                )
            )
        commons_config["flows"] = dfcx_flows
        commons_config["pages"] = dfcx_pages
        commons_config["intents"] = dfcx_intents
        commons_config["flows_map"] = dfcx_flows_map
        commons_config["pages_map"] = dfcx_pages_map
        commons_config["intents_map"] = dfcx_intents_map

        commons_config["flows_map_reverse"] = (
            self._reverse_dict(commons_config["flows_map"])
        )
        commons_config["intents_map_reverse"] = (
            self._reverse_dict(commons_config["intents_map"])
        )
        commons_config["pages_map_reverse"] = (
            {
                f_id:self._reverse_dict(pages_map) for f_id, pages_map in dfcx_pages_map.items()
            }
        )

        return commons_config`

jkshj21 commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,552 @@
+        source: dict,
+        target: dict,
+        conversation_turns: types.ConversationTurn
+        ) -> List[types.ConversationTurn]:
```
Changed the signature return type to a Union 

```
def _get_new_conversation_turns(
        self,
        source: dict,
        target: dict,
        conversation_turns: types.ConversationTurn
        ) -> None | List[types.ConversationTurn]:
```

jkshj21 replied:
```
    def _get_new_test_config(
        self,
        source: dict,
        target: dict,
        test_case: types.TestCase
        ) -> None | types.TestConfig:
```

jkshj21 commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,552 @@
+        new_conversation_turns = []
+        for conv_turn in conversation_turns:
+            if conv_turn.user_input.input.intent:
+                source_intent_id = conv_turn.user_input.input.intent
```
I break them down using the methods below:
- _get_intent_id_from_user_input
- _get_page_id_from_virtual_agent_output
- _get_intent_id_from_virtual_agent_output

```
        new_conversation_turns = []
        for conv_turn in conversation_turns:
            user_input_intent = conv_turn.user_input
            virtual_agent_output = conv_turn.virtual_agent_output

            if user_input_intent.input.intent:
                conv_turn.user_input.input.intent = self._get_intent_id_from_user_input(
                    user_input = user_input_intent,
                    source = source,
                    target = target
                )

                if not conv_turn.user_input.input.intent:
                    return None

            if virtual_agent_output.current_page:
                conv_turn.virtual_agent_output.current_page.name = (
                    self._get_page_id_from_virtual_agent_output(
                        virtual_agent_output = virtual_agent_output,
                        source = source,
                        target = target
                    )
                )

                if not conv_turn.virtual_agent_output.current_page.name:
                    return None

            if virtual_agent_output.triggered_intent:
                conv_turn.virtual_agent_output.triggered_intent.name = (
                    self._get_intent_id_from_virtual_agent_output(
                        virtual_agent_output = virtual_agent_output,
                        target = target
                    )
                )

                if not conv_turn.virtual_agent_output.triggered_intent.name:
                    return None

            new_conversation_turns.append(conv_turn)
```

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+
+        source_intent_id = user_input.input.intent
+        source_intent = self._get_intent_name_by_intent_id(
+            intent_id = source_intent_id,
+            intents_map = source["intents_map"])
```
Same as line 229, 230

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+            intent_id = source_intent_id,
+            intents_map = source["intents_map"])
+        target_intent_id = self._get_intent_id_by_intent_name(
+            intent_name = source_intent,
+            intents_map = target["intents_map_reverse"])
```
Same as line 229, 230

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+
+    def _get_new_conversation_turns(
+        self,
+        source: dict,
+        target: dict,
```
Since these two represent agent ids, use `source_agent_id` and `target_agent_id` instead

jkshj21 replied:
oh they are not agent ids. they are commons_config dictionaries. Fixed the comments

kmaphoenix replied:
Let's update the args to reflect something more semantically meaningful like `source_resource_map`, `target_resource_map`.  

This is in conjunction with my other comment about not using the word `common` as it's quite vague.

kmaphoenix replied:
Also, let's update the name of this method to `_build_conversation_turns_for_target_agent` to be more explicit about what this is doing.  We're not actually getting conversation turns here.

jkshj21 replied:
- Changed the name of the function to "_build_conversation_turns_for_target_agent"
- Instead of commons_config, changed the name to "resource_map"

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+            if user_input_intent.input.intent:
+                conv_turn.user_input.input.intent = (
+                    self._get_intent_id_from_user_input(
+                        user_input = user_input_intent,
+                        source = source,
+                        target = target
```
Same as line 229, 230 + you can put these in one line to save some space

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+            if virtual_agent_output.current_page:
+                conv_turn.virtual_agent_output.current_page.name = (
+                    self._get_page_id_from_virtual_agent_output(
+                        virtual_agent_output = virtual_agent_output,
+                        source = source,
+                        target = target
```
Same as 445 - 447

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+            if virtual_agent_output.triggered_intent:
+                conv_turn.virtual_agent_output.triggered_intent.name = (
+                    self._get_intent_id_from_virtual_agent_output(
+                        virtual_agent_output = virtual_agent_output,
+                        target = target
```
Same as line 229, 230

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+                s_commons, t_commons, tc.test_case_conversation_turns)
+            new_last_test_conv_turns = self._get_new_conversation_turns(
+                s_commons, t_commons, tc.last_test_result.conversation_turns)
+            if None in [
```
you can use `if not all([...])` instead

jkshj21 replied:
changed to 
""if not all[new_test_config, new_conv_turns, new_last_test_conv_turns]""

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+                new_test_config,
+                new_conv_turns,
+                new_last_test_conv_turns]:
+                if not new_test_config:
```
using `if new_test_config is None` is a safer option here
Exmaple: https://github.com/GoogleCloudPlatform/dfcx-scrapi/pull/166

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+                new_last_test_conv_turns]:
+                if not new_test_config:
+                    logging.warning(
+                        f"test_case: {tc.display_name} "\
+                        f"Reason: test_config is None.")
```
You don't need `\` at the end. Strings will concatenate anyway and also you don't need the second string to use f string formats.
```
                    logging.warning(
                        f"test_case: {tc.display_name} "
                        "Reason: test_config is None.")
```
with that being said, for logging messages you need to use [this format](https://google.github.io/styleguide/pyguide.html#3101-logging)

jkshj21 replied:
ok, fixed

MRyderOC commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,559 @@
+                    logging.warning(
+                        f"test_case: {tc.display_name} "\
+                        f"Reason: test_config is None.")
+                elif not new_conv_turns:
+                    logging.warning(
+                        f"test_case: {tc.display_name} "\
+                        f"Reason: test_conversation_turns is None.")
+                elif not new_last_test_conv_turns:
+                    logging.warning(
+                        f"test_case: {tc.display_name} "\
+                        f"Reason: last_test_result is None.")
```
Same as the two comments above

jkshj21 replied:
ok, sounds good - I'm working on it. Thank you!

MRyderOC reviewed (commented):
This looks good in general. Most of my comments are nit picky and related to style guide.

There are a couple of methods that you can get rid of since they are pretty straight forward such as `_get_flow_display_name_by_id`, `_get_intent_name_by_intent_id`, etc. Usually you don't need extra methods to use `.get` for dictionaries. However a method like `_get_page_id_by_page_name` is more than just `.get` and is easier to work with.

As we discussed internally, try to implement a method that migrate just a single test case which gives the user more flexibility for migrating test cases. I would suggest to name this method `migrate_test_case` and change the name of the `migrate_test_cases` to `migrate_all_test_cases`.

jkshj21 commented:
Instead of migrating all the test cases in the source agent, the `convert_test_case` function converts a single test case from the source to the target agent. I think "convert" makes more sense since we are working on a single test case. 

`self.source_commons` and `self.target_commons` are instantiated in `__init__` function
```
class TestCasesUtil(scrapi_base.ScrapiBase):
    """Util class for CX test cases"""

    def __init__(
        self,
        creds_path: str = None,
        creds_dict: Dict = None,
        creds=None,
        scope=False,
        agent_id: str = None,
    ):
        super().__init__(
            creds_path=creds_path,
            creds_dict=creds_dict,
            creds=creds,
            scope=scope,
        )

        if not agent_id:
            self.agent_id = agent_id
        self.source_commons = {}
        self.target_commons = {}
```

Every time, the users call this function, `convert_test_case`, it validates if the instance has `self.source_commons `AND `self.target_commons ` or not to prevent exhausting the resources. if not, then it sets the `commons_config` dictionaries 
```
        if not self.source_commons.get("agent_id") == source_agent_id:
            self.source_commons = self._get_commons_config(
                agent_id=source_agent_id)
        if not self.target_commons.get("agent_id") == target_agent_id:
            self.target_commons = self._get_commons_config(
                agent_id=target_agent_id)
```

I think the below wouldn't work because we want to ensure it uses the same agent id as well
`if not self.source_commons:`
therefore, I used this method instead 
`if not self.source_commons.get("agent_id") == source_agent_id:`

I added a clear note about types.testcase must have the conversation_turns in order to convert

```
This method converts the test case from the source agent to the target agent.
Note: In dfcx_scrapi.core.test_case.TestCases, it requires list_test_cases(include_conversation_turns=True),
Otherwise, it will fail to convert.
```

Added try/catch to indicate the pass and fail cases.
```
        try:
            new_test_case = self.target_commons["test_cases"].create_test_case(
                test_case=test_case)
            time.sleep(rate_limit)
        except core_exceptions.InternalServerError as err:
            logging.error(
                f"-- ERROR -- InternalServerError caught on CX.detect -- {err}")
            logging.error("test_case: %s", test_case.display_name)
            return None
        except core_exceptions.ClientError as err:
            logging.error(
                f"-- ERROR -- ClientError caught on CX.detect -- {err}")
            logging.error("test_case: %s", test_case.display_name)
            return None

        logging.info(
            "-- SUCCESS -- DFCX test_case converted -- "
            f"test case: {new_test_case.display_name}")
        return new_test_case
```

I thoroughly tested the script against the various failing scenarios below:
- When the target agent doesn't have the same intent in conversation_turn, last_test_result.conversation_turns
- when the target agent doesn't have the same page in conversation_turn and test_config, last_test_result.conversation_turns
- When the target agent doesn't have the same flow in test_config
- if the target agent has the same test_case display name

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,704 @@
+
+        return {v: k for k, v in dict_.items()}
+
+    def _get_commons_config(
```
@jkshj21 let's update the name of this method to  `_get_resource_map`
The use of `common` is vague and we want to be a bit more semantically meaningful.

jkshj21 replied:
fixed

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,704 @@
+
+        return commons_config
+
+    def _get_new_test_config(
```
@jkshj21 for this method, if both of your if statements fail (i.e. neither flow or page is found) then the method will return an empty TestConfig. Is that the intention here?

Or should it always find some value and return that config?

If it's the former, can you please update the docstring to explain this?
i.e. "If a TestConfig is found, it will return that, otherwise an empty TestConfig is returned."

jkshj21 replied:
Q: for this method, if both of your if statements fail (i.e. neither flow or page is found) then the method will return an empty TestConfig. Is that the intention here? Or should it always find some value and return that config?

Josh: Yes, that is the right intention. The reason for returning a "None" is to satisfy the below condition.
at line 516, "if not all([test_config, conv_turns, last_test_conv_turns])"

When we are converting a test_case from a source to a target agent, there are many cases when it fails. Testconfig is one of the cases that might fail. I can output an error and stop this function to run, (returns test_config or raise an error) but I thought it would be ideal if we just output an error and let the function run, because most likely, the users will convert more than one test case. 

While the test case conversion is looping, the function will stop the process if it fails then it might convert some of the test cases without looping all the way.
for test_case in list_test_cases:
    test_cases_util.convert_test_case(test_case)

I added the docstring to explain this "If a TestConfig is found, it will return that, otherwise an empty TestConfig is returned."

kmaphoenix replied:
Thanks for the detailed explanation! Updating the docstring should be sufficient.

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,704 @@
+
+        return target_test_config
+
+    def _get_flow_id_from_test_config(
```
@jkshj21 this is redundant code that is [already implemented in the TestCases ](https://github.com/GoogleCloudPlatform/dfcx-scrapi/blob/main/src/dfcx_scrapi/core/test_cases.py#L81) class.

Instead of reimplementing here, just access it from `self._global_test_cases` instead.

jkshj21 replied:
Removed and it uses the dfcx_scrapi.core.test_cases class to call this function

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,704 @@
+            agent_id = "/".join(test_case.name.split("/")[:6])
+            return f"{agent_id}/flows/00000000-0000-0000-0000-000000000000"
+
+    def _get_page_id_from_test_config(
```
@jkshj21 same as above. access from TestCases instead.

jkshj21 replied:
Removed and it uses the dfcx_scrapi.core.test_cases class to call this function

jkshj21 replied:
@kmaphoenix - hello, when i run pylint, I'm seeing errors below. It seems like I cannot access these functions because there is a "_" at the front, which I believe should only use them as an internal or protected function. 

```
test.py:219:12: W0212: Access to a protected member _get_flow_id_from_test_config of a client class (protected-access)
test.py:223:12: W0212: Access to a protected member _get_page_id_from_test_config of a client class (protected-access)

-----------------------------------
Your code has been rated at 9.92/10
```

kmaphoenix replied:
In that case, you can modify the method names in the TestCases class and make them non-private.

i.e.
`_get_flow_id_from_test_config` -> `get_flow_id_from_test_config`

Just make sure that when you're making the modification in your IDE you use the "rename all" setting or similar to rename all the references. once you've made the update, let me know and I can also run a check.

kmaphoenix replied:
@jkshj21 actually for now, let's just ignore the lint warnings and we can fix in another branch.

Just add the following to each line:
`# pylint: disable=W0212`

kmaphoenix commented on src/dfcx_scrapi/tools/test_cases_util.py:
```diff
@@ -0,0 +1,704 @@
+
+        return new_conversation_turns
+
+    def _missing_attribute_logger(
```
@jkshj21 let's rename to `_log_missing_attributes` to adhere to the naming styles we're typically using.

i.e.
- `get_something`
- `list_something`
- `log_something`

The general style is `<action>_<thing you are performing action on>`

jkshj21 replied:
ok, understood. I keep that in mind. thank you!

kmaphoenix reviewed (approved):
LGTM

@jkshj21 as discussed offline, we'll merge this change, but let's do some clean up in follow up PRs.
